### PR TITLE
RML-Mapper optimalisation

### PR DESCRIPTION
**Pull request by imadfetouh**
[MergeRequest.pdf](https://github.com/RMLio/rmlmapper-java/files/8956248/MergeRequest.pdf)

### CODE REVIEW

File: buildNumber.properties [hunk 1]
```diff
@@ -1,3 +1,3 @@
 #maven.buildNumber.plugin properties file
-#Fri Mar 11 10:44:24 GMT 2022
-buildNumber0=361
+#Tue Jun 21 10:22:00 CEST 2022
```
Comment (DylanVanAssche) [inline buildNumber.properties:2]: Please do not bump this.

File: pom.xml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <project>
     <modelVersion>4.0.0</modelVersion>
-    <groupId>be.ugent.rml</groupId>
+    <groupId>nl.ocs.be.ugent.rml</groupId>
```
Comment (DylanVanAssche) [inline pom.xml:3]: Let's keep the groupID as it is ;)

File: src/main/java/be/ugent/rml/Executor.java [hunk 3]
```diff
@@ -288,12 +309,17 @@ private List<PredicateObjectGraph> generatePredicateObjectGraphs(Mapping mapping
 
                 //check if we are dealing with a parentTriplesMap (RefObjMap)
             } else if (pogMapping.getParentTriplesMap() != null) {
-                List<ProvenancedTerm> objects;
+                Collection<ProvenancedTerm> objects;
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/Executor.java:312]: Is this change necessary? Does it improve performance?

File: src/main/java/be/ugent/rml/Executor.java [hunk 4]
```diff
@@ -375,14 +462,9 @@ private List<ProvenancedTerm> getIRIsWithTrueCondition(Record child, Term triple
...
             Object expectedBoolean = condition.execute(recordsMap);
-
-            if (expectedBoolean instanceof Boolean) {
-                if ((boolean) expectedBoolean) {
-                    ProvenancedTerm subject = this.getSubject(triplesMap, mapping, parent, i);
-                    iris.add(subject);
-                }
-            } else {
-                logger.warn("The used condition with the Parent Triples Map does not return a boolean.");
+            if (Boolean.TRUE.equals(expectedBoolean)) {
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/Executor.java:465]: This is faster than the original code?

File: src/main/java/be/ugent/rml/MappingFactory.java [hunk 5]
```diff
@@ -266,11 +266,11 @@ private void parseObjectMapWithCallback(Term objectmap, BiConsumer<MappingInfo,
                         Map<String, Object[]> parameters = new HashMap<>();
 
                         boolean ignoreDoubleQuotesInParent = this.areDoubleQuotesIgnored(store, parentTriplesMap);
-                        SingleRecordFunctionExecutor parent = new ReferenceExtractor(parents.get(0), ignoreDoubleQuotesInParent);
+                        SingleRecordFunctionExecutor parent = new ReferenceExtractor(parents.get(0).toUpperCase(), ignoreDoubleQuotesInParent);
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/MappingFactory.java:269]: I see everywhere this `toUpperCase` be added, is there a specific reason for this?

File: src/main/java/be/ugent/rml/access/AccessFactory.java [hunk 6]
```diff
@@ -259,10 +259,16 @@ private RDBAccess getRDBAccess(QuadStore rmlStore, Term source, Term logicalSour
             password = passwordObject.get(0).getValue();
         }
 
+        String connectionParameters = "";
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/access/AccessFactory.java:262]: Ah interesting! Moving out the connectionParameters helps for performance?

File: src/main/java/be/ugent/rml/extractor/ReferenceExtractor.java [hunk 7]
```diff
@@ -22,13 +25,7 @@ public ReferenceExtractor(String reference) {
 
     @Override
     public List<Object> extract(Record record) {
-        String temp = this.reference;
-
-        if (ignoreDoubleQuotes && temp.startsWith("\"") && temp.endsWith("\"")) {
-            temp = temp.substring(1, temp.length() - 1);
-        }
```
Comment (DylanVanAssche) [inline src/main/java/be/ugent/rml/extractor/ReferenceExtractor.java:29]: Ah you moved this problem to the constructor so it only happens once?

**Review by DylanVanAssche — CHANGES_REQUESTED** [review]
Hey! Thanks for your pull request! We really appreciate that you found some bottlenecks!
To get this merged, we have a few questions & comments, would you mind to have a look?

File: src/main/java/be/ugent/rml/MappingFactory.java (hunk 5 above)
Comment (klafbang) [reply src/main/java/be/ugent/rml/MappingFactory.java:269]: Yeah, as you note below, the old code did case-insensitive comparisons on each data lookup (*many* times), new code does it on creation.  Yields around a factor 5-10 in speed on our test data.

File: src/main/java/be/ugent/rml/records/CSVRecord.java [hunk 8]
```diff
@@ -56,25 +56,32 @@ public String getDataType(String value) {
...
     public List<Object> get(String value) {
-        String toDatabaseCase;
-        if (this.data.containsKey(value.toUpperCase())) {
-            toDatabaseCase = value.toUpperCase();
-        } else if (this.data.containsKey(value.toLowerCase())) {
-            toDatabaseCase = value.toLowerCase();
-        } else {
-            toDatabaseCase = value;
-        }
-        if(!this.data.containsKey(toDatabaseCase)){
```
[the remark this answers is not in the document]
Comment (klafbang) [reply src/main/java/be/ugent/rml/records/CSVRecord.java:67]: The comparison is done on creation instead of each time now.  IT has only been tested on MySQL.

File: src/main/java/be/ugent/rml/Executor.java (hunk 3 above)
Comment (klafbang) [reply src/main/java/be/ugent/rml/Executor.java:312]: It is.

getIRIsWithSimpleJoin (new) handles a special case of joins much more efficiently (instead of comparing all pairs of tuples on a join using a predicate, it uses lookup with a map, making joins in the most common case – join on a single untransformed column in both tables – linear instead of quadratic).  getIRIsWithSimpleJoin returns a collection instead of a list to avoid copying the fields.

Yields linear improvement; in our cases, up to factor 1000-10000 for realistic data examples.

Since the use of objects does not depend on list-specific operations (random lookup) there is no downside to this even though the cost could be reduced in a preprocessing step.

File: pom.xml (hunk 2 above)
Comment (klafbang) [reply pom.xml:3]: LOL, oops.  Seems the internal name we used during development accidentally dot into the PR :)

File: src/main/java/be/ugent/rml/access/AccessFactory.java (hunk 6 above)
Comment (klafbang) [reply src/main/java/be/ugent/rml/access/AccessFactory.java:262]: It does not – it was moved out do add a bit more felxibility instead of hardcoding modifications to the parameters.  Should probably have been a separate PR.

File: src/main/java/be/ugent/rml/extractor/ReferenceExtractor.java (hunk 7 above)
Comment (klafbang) [reply src/main/java/be/ugent/rml/extractor/ReferenceExtractor.java:29]: Yeah, same as with to UpperCase.

File: src/main/java/be/ugent/rml/Executor.java (hunk 4 above)
Comment (klafbang) [reply src/main/java/be/ugent/rml/Executor.java:465]: It is (or it can be depending on JRE).  The original did an instanceof check, a cast (= another instanceof check), a method call to a field access, and a boolean evaluation.  The JIT can inline the method call and likely reduce to just one instanceof check.

The optimized code is a method call to an object identity check plus a boolean evaluation.  This works because there will only ever be two instances of Boolean (TRUE + FALSE), allowing the object identity check to suffice (((Boolean) true) == ((Boolean) true)).  This effectively optimizes the instanceof check away, and only has impact on code called very often.

I don't recall the actual performance improvement.  It's less than anything else, especially after the use of the more efficient join significantly reduces how often this code is called.

**Comment by imadfetouh** [comment]
Hello,

Sorry for the late response to this pull request. Some things came up that caused this to be postponed.
I have updated my version to 6.0.0. I also tested the performance improvements again to make sure it's just as fast in version 6.0.0. 
For this version, I removed the changes about "Access". I will do this in a separate PR, so this PR is only about performance improvements.

Other changes I did in this version:
- buildNumber.properties is restored
- pom.xml is restored

Feel free to ask me some questions about this PR.

**Comment by DylanVanAssche** [comment]
Superseeded by #212
